### DISCUSSION TITLE
Sybase NHibernate Core - ASE Native driver or DataAction

### DISCUSSION THREAD

**Discussion by thomasjbyrneire** [nhibernate/nhibernate-core#2741] [General]
Does NHibernate Core work with Sybase (SybaseAdoNet4Driver) - I assume it doesn't because the Sybase.AdoNet4.AseClient driver does not seem to support .Net core.
I found a DataAction Ado driver (AdoNetCore.AseClient - https://github.com/DataAction/AdoNetCore.AseClient) which can connect .Net Core to Sybase 15.7.. BUT is there any way to get this driver to work with NHibernate?

![image](https://user-images.githubusercontent.com/71854833/116279715-c6078980-a77f-11eb-9b95-5b2d82b21ded.png)

**Comment by fredericDelaporte** [comment]
Check if no one has already written and publish on Nuget a NHibernate driver for this. If it is not the case, write your own.

Writing a NHibernate driver does neither require advanced knowledge, nor many lines of code. It is just a kind of helper for NHibernate to be able to instantiate adequate connections or commands classes.

In order to avoid taking hard dependencies on DB providers, most NHibernate drivers use a reflection based approach. Since there is already a base class for this for ASE, you may just derive from it and provide it the appropriate assembly and classes names, see [`NHibernate.Driver.SybaseAseClientDriverBase`](https://github.com/nhibernate/nhibernate-core/blob/master/src/NHibernate/Driver/SybaseAseClientDriverBase.cs).

But in your case there is not much value in avoiding taking a hard dependency on the driver. It is instead the contrary: if you write a specific project for this driver, better take an explicit dependency on it, and then just directly implement `NHibernate.Driver.DriverBase`.  
This mainly means overriding two methods: `CreateConnection` and `CreateCommand`, which just have to create an instance of the adequate class, without configuring them. (NHibernate will do that.) The SqlClient driver is [an example of that](https://github.com/nhibernate/nhibernate-core/blob/07ecb9f25534d3894dfba6a3a8e8f22932bd06cf/src/NHibernate/Driver/SqlClientDriver.cs#L86), when targeting the .Net Framework.
Additionally you should likely override a few more properties like done in the `SybaseAseClientDriverBase`.
https://github.com/nhibernate/nhibernate-core/blob/07ecb9f25534d3894dfba6a3a8e8f22932bd06cf/src/NHibernate/Driver/SybaseAseClientDriverBase.cs#L45-L52
Your driver may also implement `IEmbeddedBatcherFactoryProvider` if you want to enable batching and if there is an adequate batcher for ASE. (The `NHibernate.AdoNet.GenericBatchingBatcherFactory` works likely with ASE.)
